### PR TITLE
textInput pass 3 - adding some styling for background, fix cursor \n issue, and some more cleanup

### PR DESCRIPTION
**Pull request by hanbollar**
## Linking

related to #513 

pass1 was #524, pass2 was #546

-----

all the rest of this section of the pr description is the same as in the #524 / #546 prs

- [ ] fix selection index is always zero
- [x] consolidate input handling if possible (up/down cursor special case)
- [x] handle new line wonkiness
- [x] background color niceness for visual

------------

## Required to Merge

- [ ] **PASS** - all necessary actions must pass (excluding the auto-skipped ones)
- [x] **TEST IN HEADSET** - [main dev-testing-example](https://github.com/Volumetrics-io/mrjs/tree/main/samples/index.html) and any of the other [examples](https://github.com/Volumetrics-io/mrjs/tree/main/samples/examples) still work as expected
- [x] **VIDEO** - if this pr changes something visually - post a video here of it in headset-MR and/or on desktop (depending on what it affects) for the reviewer to reference.
  - https://github.com/Volumetrics-io/mrjs/pull/578#issuecomment-2052704184
- [x] **TITLE** - make sure the pr's title is updated appropriately as it will be used to name the commit on merge
- [ ] **BREAKING CHANGE**
  - **DOCUMENTATION**: This includes any changes to html tags and their components
    - make a pr in the [documentation repo](https://github.com/Volumetrics-io/documentation) that updates the manual docs to match the breaking change
    - link the pr of the documentation repo here: *#pr*
    - that pr must be approved by `@lobau`
  - **SAMPLES/INDEX.HTML**: This includes any changes (html tags or otherwise) that must be done to our landing page submodule as an effect of this pr's updates
    - make a pr in the [mrjs landing page repo](https://github.com/Volumetrics-io/mrjs-landing) that updates the landing page to match the breaking change
    - link the pr of the landing page repo here: *#pr*
    - that pr must be approved by `@hanbollar`

### CODE REVIEW

**Comment by hanbollar** [comment]
pausing on this for a short bit

**Comment by hanbollar** [comment]
todos:

- [ ] need good default options for the inputs
- [ ] selection index might be a dangling issue after this pr

**Comment by hanbollar** [comment]
if all is good - then this is enough to re-expose this as a tag and in documentation repo with future issues created off of this for more work

**Comment by hanbollar** [comment]
> if all is good - then this is enough to re-expose this as a tag and in documentation repo with future issues created off of this for more work

future niceties include:
- [ ] text change if a password input
- [ ] red outline as something people can toggle (troika has an easy outline feature)
- [ ] css change for if the field is disabled
- [ ] ...

**Comment by michaelthatsit** [comment]
https://github.com/Volumetrics-io/mrjs/assets/19628643/2a550fcd-f3e0-435f-9fcf-09bf254d9f7b

**Comment by michaelthatsit** [comment]
Still looking at the code but the cursor seems a little higher than it should be. Also did you test with other font sizes?

**Comment by hanbollar** [comment]
> Still looking at the code but the cursor seems a little higher than it should be. Also did you test with other font sizes?

the cursor looks higher because im sort of faking the background height (the coloring one) to fit it decently with like a slight offset for niceties, but that nicety can probably be improved compared to the quick way im doing it - it matches the specific text location

and tested with other font sizes - it doesnt scale exactly as i want atm, but it does at least start at the right top and width position - was leaving that as a future noodle to not be a blocker on this one

File: src/core/entities/MRTextAreaEntity.js [hunk 1]
```diff
@@ -12,56 +12,65 @@ import { MRTextInputEntity } from 'mrjs/core/entities/MRTextInputEntity';
...
         inputElement.style.position = 'absolute';
-        inputElement.style.left = '-9999px'; // Position off-screen
+        inputElement.style.left = '-9999px'; // Position off-screen so dont have weird white spot at top
         inputElement.style.height = '1px';
         inputElement.style.width = '1px';
         inputElement.style.overflow = 'hidden';
-        document.body.appendChild(inputElement); // Ensure it's part of the DOM for event capturing
+
+        // Ensure it's part of the DOM for event capturing
+        document.body.appendChild(inputElement);
```
Comment (michaelthatsit) [inline src/core/entities/MRTextAreaEntity.js:39]: I think we talked about this but why aren't we putting this in the shadow root? might scale a little better. especially for embedded experiences.

File: src/core/entities/MRTextFieldEntity.js [hunk 2]
```diff
@@ -16,55 +14,91 @@ export class MRTextFieldEntity extends MRTextInputEntity {
      */
     constructor() {
         super();
-        this.wrapper = this.shadowRoot.appendChild(document.createElement('div'));
-        this.wrapper.innerHTML = '<slot></slot>';
+        this.lineHeight = 1.2; // Default line height, can be adjusted as needed
+        this.object3D.name = 'textField';
+
+        // this.wrapper = this.shadowRoot.appendChild(document.createElement('div'));
```
Comment (michaelthatsit) [inline src/core/entities/MRTextFieldEntity.js:20]: Related to my other comment. why a hidden field in the DOM instead of an actual field in the shadow?

File: src/core/entities/MRTextFieldEntity.js [hunk 3]
```diff
@@ -73,7 +107,7 @@ export class MRTextFieldEntity extends MRTextInputEntity {
 
         // Ensure the cursor position is updated to reflect the current caret position
         setTimeout(() => {
```
Comment (michaelthatsit) [inline src/core/entities/MRTextFieldEntity.js:109]: what's this for?

File: src/core/entities/MRTextAreaEntity.js (hunk 1 above)
Comment (hanbollar) [reply src/core/entities/MRTextAreaEntity.js:39]: oh no main reason - i can swap it to that

File: src/core/entities/MRTextFieldEntity.js (hunk 3 above)
Comment (hanbollar) [reply src/core/entities/MRTextFieldEntity.js:109]: if we end up needing to pass the event info/etc to the function within that, we want the timeout to have a delay s.t. the event actually holds all the proper information (since it's a press/trigger type one) otherwise, the events can get delayed so the next time you grab the info, you're grabbing the previous version until the next event gets called (it's like the stack gets delayed by one or more steps if that makes sense)

---

tldr atm this setup is no longer needed since i dont pass the 'event' object to the cursor function anymore, but if we ever do it again, it will be needed with something > 0

^ given this, lmk if you'd want it removed or kept just in case

File: src/core/entities/MRTextAreaEntity.js (hunk 1 above)
Comment (michaelthatsit) [reply src/core/entities/MRTextAreaEntity.js:39]: Yeah it shouldn't be a huge issue and then you wouldn't need to do the 9999999/-99999999 absolute positions. you can just set it to `display:none` in CSS.

Comment (hanbollar) [reply src/core/entities/MRTextAreaEntity.js:39]: that is so much cleaner haha

File: src/core/entities/MRTextFieldEntity.js (hunk 3 above)
Comment (michaelthatsit) [reply src/core/entities/MRTextFieldEntity.js:109]: I'd say if we're not using it we can lose it for the sake of cleanliness

File: src/core/entities/MRTextAreaEntity.js (hunk 1 above)
Comment (michaelthatsit) [reply src/core/entities/MRTextAreaEntity.js:39]: Yeah use this as an example https://github.com/Volumetrics-io/mrjs/pull/578#discussion_r1563401938

It's one of those stack overflow tricks I picked up the first time I was trying with MRTextField.

**Comment by michaelthatsit** [comment]
Overall it's pretty solid. I don't think you're blocking anything ATM so try and make the Shadow DOM changes we discussed. 

Some notes for what I think we need by v1:

- [ ] select cursor position (blocked by click events I think)
- [ ] Text selection
- [ ] copy/paste
- [ ] bonus: spell check

**Comment by michaelthatsit** [comment]
Screenshot 2024-04-12 at 5 31 19 PM

Accidental caps but just noticed that text doesn't stay in the textarea if you type too far.

**Comment by hanbollar** [comment]
> Screenshot 2024-04-12 at 5 31 19 PM
> Accidental caps but just noticed that text doesn't stay in the textarea if you type too far.

yupyup - one of the few task for this was to add in a scrolling offset - so that the number of lines are only limited to what can show up in the main area but the text can be handled properly, so i didnt make a change to limit the height of it based on text (we also cant do mask of a mask either, so that wouldnt hv been worth it)

if that's a blocker for this pr lmk, but i had added in the todos within the code as a task after this pr

**Comment by michaelthatsit** [comment]
No blocker but let's prioritize it on the next pass. Good work!
